kimchi/src/snarky/constraint_system.rs

```diff
@@ -1190,6 +1196,34 @@ impl<Field: PrimeField> SnarkyConstraintSystem<Field> {
                 ];
                 self.add_row(vars, GateType::Zero, vec![]);
             }
+            KimchiConstraint::Poseidon2 { states, last } => {
+                states
+                    .map(|round| round.map(|x| self.reduce_to_var(x)))
+                    .into_iter()
+                    .zip(self.constants.poseidon.round_constants.clone().into_iter())
+                    .chunks(5)
+                    .into_iter()
+                    .for_each(|rounds| {
+                        let (vars, coeffs) = rounds
+                            .into_iter()
+                            .flat_map(|(round, round_constants)| {
+                                round
+                                    .into_iter()
+                                    .map(Option::Some)
+                                    .zip(round_constants.into_iter())
+                            })
+                            .unzip();
+                        self.add_row(vars, GateType::Poseidon, coeffs);
+                    });
+                let last = last
+                    .map(|x| self.reduce_to_var(x))
+                    .map(Some)
+                    .into_iter()
+                    .chain(repeat(None))
+                    .take(15)
```

Comment:
IIRC you don't have to do this, you can just have a vec of size 3

Reply:
Indeed.